### PR TITLE
Refactor: Remove `timeslots.py`

### PR DESCRIPTION
**Pull request by lcapelluto**
### Summary

Update time tracking internal to `Schedule`s. No API modifications.

### Details and comments

 - `timeslots.py` replaced by `Schedule._add_timeslots`, the namedtuple `Interval`, and two functions in utils
    - Moving error message to `_add_timeslots` allows even more info in the overlap error message, and makes `Schedule.__init__` a bit cleaner
 - Schedule._timeslots is a `Dict[Channel, List[Interval]]` rather than `TimeslotCollection`
 - methods in `pulse.Instruction` and `pulse.Schedule` simplified
 - update `rescheduler` `pad` function (and fix a bug where `until` was being ignored)
 - add tests for `utils.py` functions in new `test_utils.py` file

### CODE REVIEW

File: qiskit/pulse/utils.py [hunk 1]
```diff
@@ -16,9 +16,62 @@
...
+from .exceptions import PulseError
+
 # pylint: disable=unused-argument
 
 
+Interval = namedtuple('Interval', ['start', 'stop'])
+"""Intervals [start, stop) are exclusive of the stop time."""
+
+
+def overlaps(first: Interval, second: Interval) -> bool:
```
Comment (eggerdj) [inline qiskit/pulse/utils.py]: Is this tested somewhere?

File: qiskit/pulse/commands/instruction.py [hunk 2]
```diff
@@ -65,27 +62,27 @@ def command(self):
...
     def stop_time(self) -> int:
         """Relative end time of this instruction."""
-        return self.timeslots.stop_time
+        return self.duration
 
     @property
     def duration(self) -> int:
         """Duration of this instruction."""
-        return self.timeslots.duration
+        return self._duration
```
[the remark this answers is not in the document]
Comment (lcapelluto) [reply qiskit/pulse/commands/instruction.py]: I'd thought it had never been needed, but I am wrong: there is still the `shift` instruction :( I'll have to update this

[the remark this answers is not in the document]
Comment (lcapelluto) [reply qiskit/pulse/commands/instruction.py]: Other refactors later will be able to deprecate `shift` on `Instruction` I believe

File: qiskit/pulse/reschedule.py [hunk 3]
```diff
@@ -180,19 +180,19 @@ def pad(schedule: Schedule, channels: Optional[Iterable[Channel]] = None,
...
-
-    for channel in unoccupied_channels:
-        schedule |= Delay(until)(channel)
+        if channel not in schedule.timeslots:
+            schedule |= Delay(until)(channel)
+            continue
+
+        curr_time = 0
+        for interval in schedule.timeslots[channel]:
+            if interval.start != curr_time:
```
[the remark this answers is not in the document]
Comment (lcapelluto) [reply qiskit/pulse/reschedule.py]: Just because `curr_time` is initialized to zero. We check if the next instruction starts in the future, if so, add a delay starting at `curr_time` up to `interval.start`. After that, `curr_time` is updated to the end of that next instruction, and repeat.

Looks like the existing code had a bug, which my code also has: `until` is not respected for a channel with timeslots going beyond `until`

File: qiskit/pulse/schedule.py [hunk 4]
```diff
@@ -120,23 +104,33 @@ def ch_duration(self, *channels: List[Channel]) -> int:
...
+        Return minimum start time over supplied channels. Return 0 if none of the channels
+        have been scheduled on.
 
         Args:
             *channels: Supplied channels
         """
-        return self.timeslots.ch_start_time(*channels)
+        chan_intervals = [self._timeslots[chan] for chan in channels if chan in self._timeslots]
+        if chan_intervals:
+            return min([intervals[0].start for intervals in chan_intervals])
```
[the remark this answers is not in the document]
Comment (lcapelluto) [reply qiskit/pulse/schedule.py]: yeah, timeslots are ordered (`interval[i].stop <= interval[i].start`)

[the remark this answers is not in the document]
Comment (lcapelluto) [reply qiskit/pulse/schedule.py]: It should as long as `_add_timeslots` is implemented properly

File: qiskit/pulse/schedule.py [hunk 5]
```diff
@@ -337,9 +330,9 @@ def instruction_filter(time_inst: Tuple[int, 'Instruction']) -> bool:
         def only_intervals(ranges: Iterable[Interval]) -> Callable:
```
[the remark this answers is not in the document]
Comment (lcapelluto) [reply qiskit/pulse/schedule.py]: Maybe, they are pretty tied into the method args, but those could be better documented, too. I'm going to say this is out of scope though. Hopefully we got to it during the docathon

File: qiskit/pulse/schedule.py [hunk 6]
```diff
@@ -410,6 +403,45 @@ def draw(self, dt: float = 1, style: Optional['SchedStyle'] = None,
...
+        for channel in schedule.channels:
+            channel_intervals = [Interval(start=i.start + time, stop=i.stop + time)
+                                 for i in schedule._timeslots[channel]]
+
+            if channel not in self._timeslots:
+                self._timeslots[channel] = channel_intervals
+                continue
+
+            for idx, interval in enumerate(channel_intervals):
+                if interval.start >= self._timeslots[channel][-1].stop:
```
[the remark this answers is not in the document]
Comment (lcapelluto) [reply qiskit/pulse/schedule.py]: This should be the only method where `_timeslots` is modified, and it sorts the new timeslots being added, using the simple binary search `insertion_index` function. Would welcome close inspection of that function

File: qiskit/pulse/utils.py [hunk 7]
```diff
@@ -16,9 +16,62 @@
...
+    """
+    if first.start == second.start == second.stop:
+        # They fail to overlap if one of the intervals has duration 0
+        return False
+    if first.start > second.start:
+        first, second = second, first
+    return second.start < first.stop
+
+
+def insertion_index(intervals: List[Interval], new_interval: Interval, index: int = 0) -> int:
```
[the remark this answers is not in the document]
Comment (lcapelluto) [reply qiskit/pulse/utils.py]: Not explicitly -- any time instructions are added to a schedule, this is implicitly tested. I was lax about it before because there are many existing tests like this. I agree that these should be tested specifically; I'll add the tests

File: qiskit/pulse/commands/instruction.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (lcapelluto) [reply qiskit/pulse/commands/instruction.py]: Oh, just kidding, shift returns a `Schedule`, so this is correct.

[the remark this answers is not in the document]
Comment (lcapelluto) [reply qiskit/pulse/commands/instruction.py]: I think there are methods that can be deprecated, but I don't want to do it in this PR.

**Comment by itoko** [comment]
Thank you for creating this PR, @lcapelluto 
I don't disagree with the removal of Timeslot class, I think it's a right direction, but I have another opinion how to do it.

I think we should keep TimeslotCollection (may be renamed to Timeslots?) and refactor it as you refactored within Schedule, i.e. it has a dictionary of Interval instead of Timeslot.
Keeping TimeslotCollection makes the refactoring (without changing Instruction/Schedule) in the future when we have performance issues in Schedule. TimeslotCollection is introduced to separate data structure from Schedule interface.
For example, currently `shift` takes linear time to the size of Schedule.
If we keep TimeslotCollection and make it has `offset_time`, `shift` can be done in constant time.
(To do that, we need to modify `merge` to take the `offset_time` into consideration.)

Anyway, I like the idea "more info in the overlap error message". I think we can think of the way to provide "more info" with keeping TimeslotCollection. 
What do you think?

**Comment by lcapelluto** [comment]
@itoko I think the logic here is too simple to warrant a specialized class. This doesn't prevent us from improving the runtime of `shift` other ways -- but we will be moving towards relative timing, so a shift instruction would just require prepending a Delay.

There is a lot still we can do to improve runtime in multiple places. If we make changes again in the future, and a class seems warranted, it can be added again later. All these changes are abstracted from the user. No API changes means we have a bit of freedom here

**Comment by itoko** [comment]
@lcapelluto I missed moving towards relative timing, which will improve the runtime of `shift`.
I understand there is no strong need to keep a specialized class (TimeslotCollection).
I have no concern on the approach now. Thank you for clarification. 

I'm looking through the details of code changed. I'll leave some comments inline later.

File: qiskit/pulse/reschedule.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (itoko) [reply qiskit/pulse/reschedule.py]: Ah, I thought `interval.start > curr_time` but it's equivalent with `interval.start != curr_time` because `interval.start >= curr_time == previous_interval.end` is implicitly guranteed.

File: qiskit/pulse/utils.py [hunk 8]
```diff
@@ -16,9 +16,62 @@
...
+
+
+def overlaps(first: Interval, second: Interval) -> bool:
+    """
+    Return True iff first and second overlap.
+
+    Note: first.stop may equal second.start, since Interval stop times are exclusive.
+    """
+    if first.start == second.start == second.stop:
+        # They fail to overlap if one of the intervals has duration 0
```
Comment (itoko) [inline qiskit/pulse/utils.py]: Ah, `first.start == first.end` case is covered in Line 43.. I think there is no bug here, but how about adding tests including zero duration intervals like below for safety in the future?
```
# For some integers a, b (a < b)
assertFalse(has_overlap(Interval(a,a), Interval(a,a)))
assertFalse(has_overlap(Interval(a,a), Interval(a,b)))
assertFalse(has_overlap(Interval(a,b), Interval(a,a)))
assertFalse(has_overlap(Interval(a,b), Interval(b,b)))
assertFalse(has_overlap(Interval(b,b), Interval(a,b)))
assertTrue(has_overlap(Interval(a,a+2), Interval(a+1,a+1)))
assertTrue(has_overlap(Interval(a+1,a+1), Interval(a,a+2)))
```

File: qiskit/pulse/utils.py [hunk 9]
```diff
@@ -16,9 +16,62 @@
...
+        index: A running tally of the index, for recursion. The user should not pass a value.
+
+    Returns:
+        The index into intervals that new_interval should be inserted to maintain a sorted list
+        of intervals.
+
+    Raises:
+        PulseError: If new_interval overlaps with the given intervals.
+    """
+    if len(intervals) == 1:
```
Comment (itoko) [inline qiskit/pulse/utils.py]: Don't we need to care about empty intervals case? `if not intervals: return 0`?

**Comment by taalexander** [comment]
Can you please help me understand what problem this is solving? As far as I can tell this is offloading the problem of keeping track of time to an unstructured Python data structure composed of dictionaries and lists and attaching it to the `Schedule` and `Instruction` classes. This is a step backward from the perspective of the single responsibility principle and also violates the interface segregation principle and would make it harder to switch to say an interval representation of timeslots(not to go to hardcore on SOLID but it is very useful from a guidance perspective).

This is also not going to help with the performance issues of scheduling as these lie in the existence of many `Interval` classes that must be created when shifting pulses around. 

Even when switching to relative time having an explicit data structure where time is not associated directly with a Schedule will be useful for calculating the intersection of schedules.

In many ways, this seems like a step backward to me. Do you think we could have this discussion?

**Comment by lcapelluto** [comment]
Happy to have a discussion. The previous implementation was convoluted and obfuscated inefficiencies. Hopefully we can find a way to make a clean implementation that also pleases you. We could hold off completely until the benchmarking work is done.

**Comment by lcapelluto** [comment]
@taalexander @itoko @ajavadia 

I finally took a few hours to attempt to do some quick profiling like Itoko did for his PR. I can't seem to share the ipynb here, so I'll share in slack. Here is a summary of the results (I haven't taken statistics):

## Many instructions per channel, few channels (5)
 - (format is tested function: master -> refactor/remove-timeslots)
 - append 10,000x: 16.364 s -> 1.528 s
 - insert at sched.duration 10,000x:  64.000 s -> 1.697 s
 - insert at random time 10,000x: 17.846 s -> 1.887 s
 - shift 1,000x: 55.046 s -> 15.256 s

## Few instructions per channel (1), many channels (1,000)
 - append: 3.195 s -> 1.764 s
 - duration: 0.003 s -> 0.000 s
 - insert (random time): 14.752 s -> 9.213 s
 - shift 1,000x: 30.353 s -> 15.699 s

File: qiskit/pulse/utils.py [hunk 10]
```diff
@@ -0,0 +1,68 @@
...
+from typing import List
+
+from .exceptions import PulseError
+
+
+Interval = namedtuple('Interval', ['start', 'stop'])
+"""Intervals [start, stop) are exclusive of the stop time."""
+
+
+def overlaps(first: Interval, second: Interval) -> bool:
```
Comment (taalexander) [inline qiskit/pulse/utils.py]: Why is this not part of the interval object?

File: qiskit/pulse/utils.py (hunk 9 above)
Comment (taalexander) [reply qiskit/pulse/utils.py]: Yes, I agree with @itoko, I think this is an edge case that needs to be handled with a test.

**Comment by lcapelluto** [comment]
No clear way to just rebase or merge because files moved and altered. Reopening: https://github.com/Qiskit/qiskit-terra/pull/4147
